docs(srs): add relevant facts and operational requirements (#83)
aarnphm/tinymorph: docs/content/SRS/SRS.md

### 13.3 Requirements for Interfacing with Adjacent Systems

- **API Interfacing**: Tinymorph should integrate with the external APIs, and here is OpenAI's API. Tinymorph should also be able to handle data sources, getting content from databases.
- **System Interactions**: Tinymorph's editor may need to integrate with word processors and cloud storage systems to export and save their work. 

### 13.4 Productization Requirements

- **Packaging and Distribution**: Tinymorph should be packaged with installers or package managers in a way that easy to distribute across platforms and easy to to updates. 
- **Documentation**: Clear user guide and technical document should be accompanied with the Tinymorph software, easy to access.

### 13.5 Release Requirements

- **Versioning**: Each time the release of tinymorph must be clearly versioned with changelogs outlining new features, bug fixes, and improvements. All modification should also be recorded in relevant documents. 

## 14. Maintainability and Support Requirements

### 14.1 Maintenance Requirements

- **Regular Updates**: Regular updates, including adding features and addressing bugs, should be provided during its life span. 

> - **Backward Compatibility**: User's personalised preferences and setups should be kept and compatible with the new updates and maintenance. 

Reviewer: also no need to mention backward compat for now.

> [!IMPORTANT] OER-WER1
>
> Tinymorph should be able to handle different internet speed for host inference.

Rationale: If user choose to use a host inference, a stable network connection is needed. Thus tinymorph need to adjust to different internet speed to provide a dependable performance, and give notice if network condition is badly concerned. 

### 13.3 Requirements for Interfacing with Adjacent Systems

> [!IMPORTANT] OER-RIAS1
>
> Tinymorph should have a good intergration with different model.

Rationale: Since tinymorph implements SAEs training over the Open-AI existing model, there need to be agood integration between these two parts.

> [!IMPORTANT] OER-RIAS2
>
> Tinymorph should cooperate with cloud model running well.

Rationale: For hosted inference, since the model is running remotely, the integration between model running and information return need better organized to make the process soomth.

### 13.4 Productization Requirements

> [!IMPORTANT] OER-PR1
>
> Tinymorph have a easy implementation on user side

Rationale: Tinymorph should be easy to implement for user end. Updates and distribution on the user end should also cost little effort.

> > [!IMPORTANT] OER-PR1
> >
> > Updated relevant documentation should be easy to find for user
> 
> Rationale: The manual and technical document should be accessed by user without any barrier. Accompanied with the software, corresponding updates should also implemented on documents.

Reviewer: ```suggestion
> [!IMPORTANT] OER-PR2
>
> Relevant documentation should be accessible by users.

Rationale: Usage manuals and technical-related details should be easily accessible from `tinymorph`'s interface.

> [!IMPORTANT] OER-PR3
>
> Feedback should also be included within the interface

Rationale: Enable user-feedback to improve the product.
```

### 13.5 Release Requirements

> > [!IMPORTANT] OER-RR1
> >
> > Version controls and updated information about new release should be recorded
> 
> Rationale: There should be a platform or document used to demonstrate the relevant version information and new release updates, like bug fixes and improvements. 

Reviewer: ```suggestion
> [!IMPORTANT] OER-RR1
>
> Release cycle must utilize current GitHub CD workflow.

Rationale: Version control and release cycle should follow semantic versioning and utilize GitHub's CI for automation.

> [!IMPORTANT] OER-RR2
>
> End-to-end tests should pass before deploying to production.

Rationale: end-to-end workflow must be the minimum for all feature development to ensure `tinymorph` is functional within a production environment.
```

## 14. Maintainability and Support Requirements

### 14.1 Maintenance Requirements

> > [!IMPORTANT] OER-RR1
> >
> > Updates should be provided during tinymorph's life span
> 
> Rationale: Regular updates, including adding features and addressing bugs on user end, and iteration on model should be carried out during the tinymorph's life span as maintenance.

Reviewer: ```suggestion
> [!IMPORTANT] OER-MR1
>
> Security updates must be done periodically

Rationale: Regular security updates to adjacent dependencies must be done quickly to avoid certain CVE exploits if they exist.

> [!IMPORTANT] OER-MR2
>
> Feature integrations must pass existing tests

Rationale: Given features works must not fail existing testing infrastructure.
```

### 14.2 Supportability Requirements

> > [!IMPORTANT] OER-SR1
> >
> > Feedback from user should be able to gathered and pass to developers
> 
> Rationale: There should be a platform used to gather and record the feedback or report for bugs from end user as supplement for future development. 

Reviewer: ```suggestion
> [!IMPORTANT] OER-SR1
>
> User feedback loop must be present.

Rationale: For further development and UX improvement, a user-feedback loop is required. 
```

### 14.3 Adaptability Requirements

> > [!IMPORTANT] OER-AR1
> >
> > Tinymorph should adapt to main stream OS and devices as running environment
> 
> Rationale: Tinymorph should be able to provide services over the common OS including macOS, Windows and linux, and user end should perform well on average hardware environment. 

Reviewer: ```suggestion
> [!IMPORTANT] OER-AR1
>
> `tinymorph` must be able to run with existing users' environment

Rationale: For web interface, `tinymorph` should be able to run on all existing modern browser. For packaged binary, it must support major architectures and operating system. 
```

### 16.1 Cultural Requirements

> [!IMPORTANT] CulR-CR1
>

> > Tinymorph support English input and output
> 
> Rationale: In the first edition of tinymorph development, it only support for English input content and can only generate text in English.

Reviewer: ```suggestion
> English supports

Rationale: English will be supported for alpha release of `tinymorph`. This is due to the limited capabilities of models when dealing with multilingual inputs.
```

> [!IMPORTANT] CulR-CR2
>

> > Culturally inappropriate content is filtered out from output language
> 
> Rationale: If there are any inappropriate content specific to the culture of generated output language, they should be filtered out.

Reviewer: ```suggestion
> Cultural reference must be factual

Rationale: If given cultural references are generated, it must utilize tools to fact-check given suggestions (using web-search).
```

> > [!IMPORTANT] CulR-CR3
> >
> > Using left-to-right (LTR) generating habit
> 
> Rationale: Currently only LTR habit implemented in the tinymorph to present output. 

Reviewer: ```suggestion
> [!IMPORTANT] CulR-CR3
>
> Support left-to-right (LTR) reading flow

Rationale: Panels will be presented in LTR manner. RTL will be supported once multilingual are added. 
```

## 17. Compliance Requirements

### 17.1 Legal Requirements

> [!IMPORTANT] CompR-LR1
>

> > The generated contents is not protectable under US copyright law
> 
> Rationale: Since the genereted content invloves Open-AI model output, the generated content is subject to OpenAI's license and terms of use, which is not protectable under US copyright law

Reviewer: ```suggestion
> Suggestion must follow strict US copyright law.

Rationale: Certain suggestions, if any, uses additional contents, references must be made to ensure copyright law compliance.

> [!IMPORTANT] CompR-LR2
>
> [SOC2](https://www.vanta.com/products/soc-2) compliance

Rationale: `tinymorph` should follow SOC-2 attestation for its inference server.
```

> [!IMPORTANT] CompR-LR2
>

> > User agrees to send necessary data to host server for return from language model
> 
> Rationale: To use the language model in host server, user need to send necessary information based on input text to trigger language model. Users are consent for this data collection.

Reviewer: ```suggestion
> Users permission to run inference against their content

Rationale: `tinymorph` will require users' inputs to make corresponding suggestions. In other words, existing user content will be sent during inference requests.
```

### 17.2 Standards Compliance Requirements

> [!IMPORTANT] CompR-SCR1
>

> > Tinymorph align with standard HTTP protocol
> 
> Rationale: Tinymorph adhere to Hypertext Transfer Protocol (HTTP/1.1 and HTTP/2) standards as defined by the Internet Engineering Task Force (IETF) in RFC 2616 (for HTTP/1.1) and RFC 7540 (for HTTP/2).

Reviewer: ```suggestion
> follows standard HTTP protocol for client-server communication

Rationale: `tinymorph` will adhere to Hypertext Transfer Protocol (HTTP/1.1) standards as defined by the Internet Engineering Task Force (IETF) in RFC 2616 (for HTTP/1.1).
```

### 22.1 Requirements for Migration to the New Product

> [!IMPORTANT] MNP-RMNP1
>

> > During upgrading of cloud model, user end should have access to host server as usual
> 
> Rationale: Whenever there is a iteration on the model running on host server, it should make no effect to end user's generating process. An updating strategy that change the model when no user is in use should be implemented.

Reviewer: ```suggestion
> Minimal downtime during migration process

Rationale: When the inference server is updating or maintaining, users should be aware of the downtime, given it shouldn't affect users' workflow.
```

> > [!IMPORTANT] MNP-RMNP1
> >
> > A notice is given to user if new edition of UI is available
> 
> Rationale: If there is a new edition of UI is distributed, a notice on current UI should be provided for user to access to latest edition. 

Reviewer: ```suggestion
```

### 22.2 Data That Has to be Modified or Translated for the New System

> > [!IMPORTANT] MNP-DMTNS1
> >
> > Updated user interface should update user local stored preferences in readable data format
> 
> Rationale: Whenever there is a iteration on the user end interface, tinymorph need to make local stored user preferences readable and in up-to-date format. 

Reviewer: ```suggestion
> [!IMPORTANT] MNP-DMTNS1
>
> Migration in future config format should ensure backward compatibility for one-time transition

Rationale: When configuration or certain features require breaking change, `tinymorph` must be able to migrate existing configuration to the new format without breaking change. 
```
